### ISSUE TITLE
two Distributed tables  join extremely slow and  the initial query has huge amount of threads 

### ISSUE THREAD

**Issue by windylcx** [ClickHouse/ClickHouse#49000]
My CK version is 22.8.2.11 ,
and the the cluster has 63 nodes  so one table has 63 shards and 2 replicas for each shard.
and i try to find where some distributed join so much slow (less than 103.07 KB/s./s)

in another test cluster  ,it can't finish the query with the exception 
DB::Exception: Can't initialize pipeline with empty pipe.: While executing Remote.

table a :has more than  40000000 rows 
table b :has   more than 1100000 rows 

**here's the process info:**
it has more than 5k threads

query:                     create table test.tt4 engine=MergeTree() order by uid as  select
        a.aid as aid
        ,b.bid as bid
        ,a.uid as uid
        ,a.type as type
        ,a.gid as gid
        ,a.name as name
        ,a.m as m
        ,a.terminal as terminal
        ,a.os as os
        ,a.app as app
        from test.t1_dist a
        global left  join test.t2_dist b on b.bid=a.aid
        Settings send_logs_level='trace';
thread_ids:           [45810,104646,104716,103496,103685,169827,170053,104548,103840,104302,103225,104332,103581,104120,26649,
...................................
26696,27452,27622,28387,24079,27130,28387,27098,27733,28177,28372,27733,26689,26755,28332,27076,26755,28332,27622,28356,24079,28387,27130,27098]
ProfileEvents:        {'Query':1,'FileOpen':833,'ReadBufferFromFileDescriptorReadBytes':36481337,'WriteBufferFromFileDescriptorWrite':2,'WriteBufferFromFileDescriptorWriteBytes':346,'FileSync':2,'FileSyncElapsedMicroseconds':6356,'ReadCompressedBytes':265347914,'CompressedReadBufferBlocks':49942,'CompressedReadBufferBytes':535611926,'OpenedFileCacheMisses':831,'IOBufferAllocs':1850,'IOBufferAllocBytes':157827336,'ArenaAllocChunks':435,'ArenaAllocBytes':35278848,'MarkCacheHits':831,'CreatedReadBufferOrdinary':831,'DiskReadElapsedMicroseconds':44835,'DiskWriteElapsedMicroseconds':60,'NetworkReceiveElapsedMicroseconds':472865,'NetworkSendElapsedMicroseconds':21586646022,'NetworkReceiveBytes':251571455,'NetworkSendBytes':4505296247,'InsertedRows':648307,'InsertedBytes':98139420,'SelectedParts':831,'SelectedRanges':831,'SelectedMarks':831,'SelectedRows':2943008,'SelectedBytes':900485002,'ContextLock':285,'RWLockAcquiredReadLocks':30,'RealTimeMicroseconds':255105525,'UserTimeMicroseconds':21550988,'SystemTimeMicroseconds':2490990,'SoftPageFaults':467456,'QueryProfilerRuns':30384,'ThreadPoolReaderPageCacheMiss':5288,'ThreadPoolReaderPageCacheMissBytes':36481337,'ThreadPoolReaderPageCacheMissElapsedMicroseconds':44835}
Settings:             {'max_insert_threads':'16','max_distributed_connections':'10000','receive_timeout':'10000','send_timeout':'10000','tcp_keep_alive_timeout':'120','idle_connection_timeout':'300','distributed_connections_pool_size':'10000','connections_with_failover_max_tries':'10','use_uncompressed_cache':'0','load_balancing':'random','aggregation_memory_efficient_merge_threads':'48','read_backoff_min_concurrency':'24','http_send_timeout':'10000','http_receive_timeout':'10000','send_logs_level':'trace'}
current_database:     default

**and each sub query only has 1 threads id**

is_initial_query:     0
user:                 xxx
query_id:             549710d8-756c-4e56-a0e8-0ba607d14fb2
address:              ::ffff:xxxx
port:                 50863
initial_user:         root
initial_query_id:     02227245-ef29-440c-9ded-077a804fb207
initial_address:      ::ffff:xxxxx
initial_port:         53097
interface:            1
os_user:              xxx
client_hostname:      rxxx
client_name:          ClickHouse
client_revision:      54460
client_version_major: 22
client_version_minor: 8
client_version_patch: 2
http_method:          0
http_user_agent:
http_referer:
forwarded_for:
quota_key:
distributed_depth:    1
elapsed:              1663.314962653
is_cancelled:         0
is_all_data_sent:     0
read_rows:            0
read_bytes:           0
total_rows_approx:    0
written_rows:         0
written_bytes:        0
memory_usage:         840211065
peak_memory_usage:    898942649
query:                SELECT `a`.`aid` AS `aid`, `b`.`bid` AS `bid`, `a`.`uid` AS `uid`, `a`.`type` AS `type`, `a`.`gid` AS `gid`, `a`.`name` AS `name`, `a`.`m` AS `m`, `a`.`terminal` AS `terminal`, `a`.`os` AS `os`, `a`.`app` AS `app` FROM `test`.`t1_local` AS `a` GLOBAL ALL LEFT JOIN `b` AS `b` ON `b`.`preOrderId` = `a`.`preOrderId`
thread_ids:           [181726]
ProfileEvents:        {'ReadCompressedBytes':142626807,'CompressedReadBufferBlocks':15567,'CompressedReadBufferBytes':279097982,'IOBufferAllocs':1,'IOBufferAllocBytes':71297,'NetworkReceiveElapsedMicroseconds':106501,'NetworkSendElapsedMicroseconds':19,'NetworkReceiveBytes':143142434,'NetworkSendBytes':789,'SelectedRows':675096,'SelectedBytes':438629920,'ContextLock':15567}
Settings:             {'max_insert_threads':'16','max_distributed_connections':'10000','receive_timeout':'10000','send_timeout':'10000','tcp_keep_alive_timeout':'120','idle_connection_timeout':'300','distributed_connections_pool_size':'10000','connections_with_failover_max_tries':'10','use_uncompressed_cache':'0','load_balancing':'random','aggregation_memory_efficient_merge_threads':'48','read_backoff_min_concurrency':'24','http_send_timeout':'10000','http_receive_timeout':'10000','send_logs_level':'trace'}
current_database:     default

**Comment by windylcx** [comment]
> 

yes,the right table has the smaller size of id, and i don't get the meaning of -s, is it a parameter?

i just want to find out why there will be so much threads in such a query,and how can i reduce it ? may be  some parameter settings like  max_threads or others?

and i found that the max_threads is auto detect while execute the query ,it depend on what ? table rows ,size or parts?

i can't find any details in the docs

**Comment by nickitat** [comment]
if you didn't change `max_threads` it will be calculated based on the number of available cpu cores.
setting it lower will lower the number of used threads, but I don't know honestly where 5k came from.

**Comment by SaltTan** [comment]
5k most likely came from #41180

**Comment by nickitat** [comment]
does the problem still exist?
